sentry-core/src/hub.rs

```diff
@@ -222,6 +224,49 @@ impl Hub {
         }
     }
 
+    /// Looks up an integration on the hub.
+    ///
+    /// Calls the given function with the requested integration instance when it
+    /// is active on the currently active client.
+    ///
+    /// # Example
+    /// ```
+    /// # use std::sync::Arc;
+    /// use sentry_core::{Client, ClientOptions, Hub, Integration};
+    ///
+    /// #[derive(Debug)]
+    /// struct MyIntegration(usize);
+    /// impl Integration for MyIntegration {}
+    ///
+    /// let client = Arc::new(Client::from(ClientOptions {
+    ///     integrations: vec![Arc::new(MyIntegration(10))],
+    ///     ..Default::default()
+    /// }));
+    /// let hub = Hub::with(|hub| Hub::new_from_top(hub));
+    /// hub.bind_client(Some(client));
+    ///
+    /// let value = hub.with_integration(|integration: &MyIntegration| integration.0);
+    /// assert_eq!(value, 10);
+    /// ```
+    pub fn with_integration<I, F, R>(&self, f: F) -> R
+    where
+        I: Integration,
+        F: FnOnce(&I) -> R,
+        R: Default,
+    {
+        with_client_impl! {{
+            if let Some(client) = self.client() {
+                let integrations = client.integrations.read().unwrap();
+                if let Some(integration) = integrations.get(&TypeId::of::<I>()) {
+                    // SAFETY: we just checked for the concrete type via TypeId
+                    let integration = unsafe { &*(&**integration as *const dyn Integration as *const I) };
```

Comment:
Since we checked, there's really no way this should fail in practice, so it would be great to avoid the `unsafe` block and just do a downcast. Gives us another `if let Some()` but at least it's safe.